qiskit/aqua/algorithms/phase_estimators/phase_estimator.py

```diff
@@ -0,0 +1,213 @@
+from qiskit import execute
+from qiskit.aqua.utils import get_subsystem_density_matrix
+import numpy as np
+
+
+class PhaseEstimator(QuantumAlgorithm):
+    """The Quantum Phase Estimation algorithm.
+    """
+
+    def __init__(self,
+                 num_evaluation_qubits,
+                 pe_circuit = None,
+                 num_unitary_qubits = None,
+                 unitary = None,
+                 input_state_circuit = None,
+                 quantum_instance: Optional[Union[QuantumInstance, BaseBackend]] = None):
+
+        """
+
+        Args:
+            num_evaluation_qubits: The number of qubits used in estimating the phase. The
+                                   phase will be estimated as a binary string with this many
+                                   bits.
+            pe_circuit: The phase estimation circuit.
+            num_unitary_qubits: Must agree with the number of qubits in the unitary in `pe_circuit` if
+                                `pe_circuit` is passed. This parameter will be set from
+                                `unitary` if `unitary` is passed.
+            unitary: The circuit representing the unitary operator whose eigenvalues (via phase) will
+                     be measured. Exactly one of `pe_circuit` and `unitary` must be passed.
+            input_state_circuit: The circuit that prepares the state whose eigenphase will be measured.
+                                 If this parameter is ommited, no preparation circuit will be run and
+                                 input state will be the all-zero state in the computational basis.
+            quantum_instance: The quantum instance on which the circuit will be run.
+        """
+
+        if unitary is None:
+            if pe_circuit is None:
+                raise ValueError('One one of `unitary` and `pe_circuit` may be `None`.')
+            else:
+                if num_unitary_qubits is None:
+                    raise ValueError('`num_unitary_qubits` is required when passing `pe_circuit`.')
+                self._pe_circuit = pe_circuit
+                self._num_unitary_qubits = num_unitary_qubits
+        else:
+            if not (num_unitary_qubits is None or num_unitary_qubits == unitary.num_qubits):
+                raise ValueError('`num_unitary_qubits` disagrees with size of `unitary`.')
+            pe_circuit = PhaseEstimation(num_evaluation_qubits, unitary)
+            self._num_unitary_qubits = unitary.num_qubits
+            self._pe_circuit = pe_circuit
+
+        self._num_evaluation_qubits = num_evaluation_qubits
+
+        if not input_state_circuit is None:
+            self._pe_circuit = pe_circuit.compose(
+                input_state_circuit,
+                qubits=range(num_evaluation_qubits, num_evaluation_qubits + self._num_unitary_qubits),
+                front=True)
+        else:
+            self._pe_circuit = pe_circuit
+
+        super().__init__(quantum_instance)
+        self._add_classical_register()
```

Comment:
Since `quantum_instance` is optional this will break if not passed, right? Only when `run` is called we can be sure that the quantum instance is set, therefore this line should be inside `_run`

Reply:
Hmm. Don't know. I'll test this. Need to remember to add unit tests with and without passing `quantum_instance`